extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -130,42 +124,55 @@ export function maybeExecuteRealTimeConfig_(a4aElement, customMacros) {
  * @param {!Object<string, boolean>} seenUrls
  * @param {!Array<!Promise<!rtcResponseDef>>} promiseArray
  * @param {number} rtcStartTime
- * @param {!Object<string, !../../../src/service/variable-source.SyncResolverDef>} macros
+ * @param {!Object<string, !../../../src/service/variable-source.AsyncResolverDef>} macros
  * @param {number} timeoutMillis
  * @param {string=} opt_vendor
  * @private
  */
 function inflateAndSendRtc_(a4aElement, url, seenUrls, promiseArray,
-  rtcStartTime, macros, timeoutMillis,
-  opt_vendor) {
+  rtcStartTime, macros, timeoutMillis, opt_vendor) {
   const win = a4aElement.win;
   const ampDoc = a4aElement.getAmpDoc();
-  if (Object.keys(seenUrls).length == MAX_RTC_CALLOUTS) {
-    return logAndAddErrorResponse_(
-        promiseArray, RTC_ERROR_ENUM.MAX_CALLOUTS_EXCEEDED,
-        opt_vendor || url);
-  }
+  /**
+   * The time that it takes to substitute the macros into the URL can vary
+   * depending on what the url requires to be substituted, i.e. a long
+   * async call. Thus, however long the URL replacement took is treated as a
+   * time penalty.
+   */
+  const send = (url, timePenalty) => {
+    if (Object.keys(seenUrls).length == MAX_RTC_CALLOUTS) {
+      return buildErrorResponse_(
+          RTC_ERROR_ENUM.MAX_CALLOUTS_EXCEEDED,
+          opt_vendor || url, undefined, true);
+    }
+    if (!isSecureUrl(url)) {
+      return buildErrorResponse_(RTC_ERROR_ENUM.INSECURE_URL,
+          opt_vendor || url, undefined, true);
+    }
+    if (seenUrls[url]) {
+      return buildErrorResponse_(RTC_ERROR_ENUM.DUPLICATE_URL,
+          opt_vendor || url, undefined, true);
+    }
+    seenUrls[url] = true;
+    if (url.length > MAX_URL_LENGTH) {
+      url = truncUrl_(url);
+    }
+    return sendRtcCallout_(
```

Comment:
Currently if the timePenalty <= 0, you still send the RTC request which seems unnecessary.  Add a check and fail immediately if exceeded.  Will also let you identify that async url expansion was the cause.

Reply:
done